crates/transaction-pool/src/pool/mod.rs

```diff
@@ -541,15 +542,25 @@ where
     }
 
     /// Adds all transactions in the iterator to the pool, returning a list of results.
+    ///
+    /// Note: A large batch may lock the pool for a long time that blocks important operations
+    /// like updating the pool on canonical state changes. The caller should consider having
+    /// a max batch size to balance transaction insertions with other updates.
     pub fn add_transactions(
         &self,
         origin: TransactionOrigin,
         transactions: impl IntoIterator<Item = TransactionValidationOutcome<T::Transaction>>,
     ) -> Vec<PoolResult<TxHash>> {
-        let mut added =
-            transactions.into_iter().map(|tx| self.add_transaction(origin, tx)).collect::<Vec<_>>();
+        let mut added = {
+            let mut pool = self.pool.write();
+            transactions
+                .into_iter()
+                .map(|tx| self.add_transaction(&mut pool, origin, tx))
+                .collect::<Vec<_>>()
+        };
 
         // If at least one transaction was added successfully, then we enforce the pool size limits.
+        // A micro-optimization is to reuse the `pool` write lock in the `added` scope above.
```

Comment:
this comment seems a bit out of place here?

can we move this into the scope above?

Reply:
Unlazied myself and implemented it right in this PR now 😅.